src/wallet/wallet_tx_builder.cpp

```diff
@@ -0,0 +1,683 @@
+                } else {
+                    if (strategy.AllowRevealedRecipients()) {
+                        if (ua.GetP2SHReceiver().has_value()) {
+                            resolvedPayments.emplace_back(
+                                ua, ua.GetP2SHReceiver().value(), payment.GetAmount(), std::nullopt, false);
+                        } else if (ua.GetP2PKHReceiver().has_value()) {
+                            resolvedPayments.emplace_back(
+                                ua, ua.GetP2PKHReceiver().value(), payment.GetAmount(), std::nullopt, false);
+                        } else {
+                            // This should only occur when we have
+                            // • an Orchard-only UA,
+                            // • `AllowRevealedRecipients`, and
+                            // • can’t resolve Orchard (which means either insufficient non-Sprout
+                            //   funds or pre-NU5).
+                            resolutionError = AddressResolutionError::CouldNotResolveReceiver;
+                        }
+                    } else if (strategy.AllowRevealedAmounts()) {
+                        resolutionError = AddressResolutionError::TransparentReceiverNotAllowed;
+                    } else {
+                        resolutionError = AddressResolutionError::RevealingReceiverAmountsNotAllowed;
+                    }
+                }
+            }
+        }, payment.GetAddress());
+
+        if (resolutionError.has_value()) {
+            return resolutionError.value();
+        }
+    }
+    auto resolved = Payments(resolvedPayments);
+
+    if (orchardOutputs > this->maxOrchardActions) {
+        return ExcessOrchardActionsError(
+                ActionSide::Output,
+                orchardOutputs,
+                this->maxOrchardActions);
+    }
+
+    // Set the dust threshold so that we can select enough inputs to avoid
+    // creating dust change amounts.
+    CAmount dustThreshold{this->DefaultDustThreshold()};
+
+    // TODO: the set of recipient pools is not quite sufficient information here; we should
+    // probably perform note selection at the same time as we're performing resolved payment
+    // construction above.
+    if (!spendableMut.LimitToAmount(targetAmount, dustThreshold, resolved.GetRecipientPools())) {
+        CAmount changeAmount{spendableMut.Total() - targetAmount};
+        return InvalidFundsError(
+                spendableMut.Total(),
+                changeAmount > 0 && changeAmount < dustThreshold
+                // TODO: we should provide the option for the caller to explicitly
+                // forego change (definitionally an amount below the dust amount)
+                // and send the extra to the recipient or the miner fee to avoid
+                // creating dust change, rather than prohibit them from sending
+                // entirely in this circumstance.
+                // (Daira disagrees, as this could leak information to the recipient)
```

Comment:
```suggestion
                // (Daira disagrees, as this could leak information to the recipient
                // or publically in the fee.)
```